nameserver/server.go

```diff
@@ -215,70 +215,49 @@ func (s *DNSServer) Stop() error {
 }
 
 func (s *DNSServer) queryHandler(proto dnsProtocol) dns.HandlerFunc {
-	return func(w dns.ResponseWriter, r *dns.Msg) {
-		now := time.Now()
-		q := r.Question[0]
-		maxLen := getMaxReplyLen(r, proto)
-		lookups := []ZoneLookup{s.Zone, s.mdnsCli}
-		Debug.Printf("Query: %+v", q)
-
-		reply, err := s.cache.Get(r, maxLen, time.Now())
-		if err != nil {
-			if err == errNoLocalReplies {
-				Debug.Printf("[dns msgid %d] Cached 'no local replies' - skipping local lookup", r.MsgHdr.Id)
-				lookups = []ZoneLookup{s.Zone}
-			} else {
-				Debug.Printf("[dns msgid %d] Error from cache: %s", r.MsgHdr.Id, err)
-				w.WriteMsg(makeDNSFailResponse(r))
-				return
-			}
-		}
-		if reply != nil {
-			Debug.Printf("[dns msgid %d] Returning reply from cache: %s/%d answers",
-				r.MsgHdr.Id, dns.RcodeToString[reply.MsgHdr.Rcode], len(reply.Answer))
-			w.WriteMsg(reply)
-			return
+	zoneLookup := func(lookup ZoneLookup, q *dns.Question, r *dns.Msg) (*dns.Msg, []ZoneRecord, error) {
+		if ips, err := lookup.LookupName(q.Name); err != nil {
+			return nil, nil, err
+		} else {
+			return makeAddressReply(r, q, ips), ips, nil
 		}
+	}
 
-		// catch unsupported queries
-		if q.Qtype != dns.TypeA {
-			Debug.Printf("[dns msgid %d] Unsuported query type %s", r.MsgHdr.Id, dns.TypeToString[q.Qtype])
-			m := makeDNSNotImplResponse(r)
-			s.cache.Put(r, m, negLocalTTL, 0, now)
-			w.WriteMsg(m)
-			return
-		}
+	fallback := func(w dns.ResponseWriter, r *dns.Msg) {
+		w.WriteMsg(makeDNSFailResponse(r))
+	}
 
-		for _, lookup := range lookups {
-			if ips, err := lookup.LookupName(q.Name); err == nil {
-				m := makeAddressReply(r, &q, ips)
-				m.Authoritative = true
+	return s.commonQueryHandler(proto, "Query", dns.TypeA, zoneLookup, fallback)
+}
 
-				Debug.Printf("[dns msgid %d] Caching response for %s-query for \"%s\": %s [code:%s]",
-					m.MsgHdr.Id, dns.TypeToString[q.Qtype], q.Name, ips, dns.RcodeToString[m.Rcode])
-				s.cache.Put(r, m, nullTTL, 0, now)
-				w.WriteMsg(m)
-				return
-			}
-			now = time.Now()
+func (s *DNSServer) rdnsHandler(proto dnsProtocol) dns.HandlerFunc {
+	zoneLookup := func(lookup ZoneLookup, q *dns.Question, r *dns.Msg) (*dns.Msg, []ZoneRecord, error) {
+		if names, err := lookup.LookupInaddr(q.Name); err != nil {
+			return nil, nil, err
+		} else {
+			return makePTRReply(r, q, names), names, nil
 		}
+	}
 
-		Info.Printf("[dns msgid %d] No results for type %s query %s [caching no-local]",
-			r.MsgHdr.Id, dns.TypeToString[q.Qtype], q.Name)
-		s.cache.Put(r, nil, negLocalTTL, CacheNoLocalReplies, now)
-		w.WriteMsg(makeDNSFailResponse(r))
+	notUsHandler := s.notUsHandler(proto)
+	fallback := func(w dns.ResponseWriter, r *dns.Msg) {
+		Info.Printf("[dns msgid %d] -> sending to fallback server", r.MsgHdr.Id)
+		notUsHandler(w, r)
```

Comment:
A picky comment: we could remove the `notUsHandler` definition and use directly `s.notUsHandler(proto)` here, right?